### PR TITLE
lots more tweaking of the 2.x docs

### PR DESCRIPTION
**Pull request by wmorgan**
- Merge "proxy injection" into "adding your service" and clean up text
- Simplify "architecture"
- Remove "next steps"
- Improve "upgrading" and "supported protocols"
- Various other tweaks

In a later branch, I plan to:
- add a "Features" page
- add the federation docs to the Prometheus docs
- maybe add an FAQ

Signed-off-by: William Morgan <william@buoyant.io>

### CODE REVIEW

File: linkerd.io/content/2/adding-your-service/_index.md [hunk 1]
```diff
@@ -6,24 +6,38 @@ title = "Adding Your Service"
...
-availability of your application will not be affected.
+In order for your service to take advantage of Linkerd, it needs to have
+Linkerd's data plane proxies installed into the service pods.
 
-To add Linkerd to your service, run:
+There are two ways to do this:
+
+1. Use `linkerd inject` ([reference](../cli/inject)) to update the YAML
+   manifest for the service, and pass that output to `kubectl`.
+2. Use [automatic proxy injection](../proxy-injection) to do this automatically
```
Comment (klingerf) [inline linkerd.io/content/2/adding-your-service/_index.md]: You're removing the `proxy-injection` page as part of this branch. This should probably link to an anchor on this page instead (`#method-2-automatic-proxy-injection).

File: linkerd.io/content/2/adding-your-service/_index.md [hunk 2]
```diff
@@ -6,24 +6,38 @@ title = "Adding Your Service"
...
+   on the cluster itself.
+
+In both cases, the proxy sidecar will be added as a separate container in the
+pod, along with an `initContainer` that configures iptables to pass all traffic
+through the proxy. Thus, no configuration or modification of the application
+itself is necessary.
+
+# Method 1: `linkerd inject`
+
+To add the Linkerd data plane proxies to a service, run:
```
Comment (klingerf) [inline linkerd.io/content/2/adding-your-service/_index.md]: "service" here makes me think of kubernetes services. And in the next paragraph we call the same target "application". So I'd prob write this as:

> To add the Linkerd data plane proxies to your application, run:

File: linkerd.io/content/2/adding-your-service/_index.md [hunk 3]
```diff
@@ -6,24 +6,38 @@ title = "Adding Your Service"
...
-`deployment.yml` is the Kubernetes config file containing your
-application. This will add the proxy sidecar along with an `initContainer` that
-configures iptables to pass all traffic through the proxy. By applying this new
-configuration via `kubectl`, a rolling update of your deployment will be
-triggered replacing each pod with a new one.
+Where `deployment.yml` is the Kubernetes config file containing your
+application.
+
+By applying this new manifest via `kubectl`, a rolling update of your
+deployment will be triggered replacing each pod with a new one.  Thanks to
```
Comment (klingerf) [inline linkerd.io/content/2/adding-your-service/_index.md]: Need a comma here: `.. will be triggered, replacing ...`

File: linkerd.io/content/2/adding-your-service/_index.md [hunk 4]
```diff
@@ -6,24 +6,38 @@ title = "Adding Your Service"
...
+Kubernetes's rolling update logic, even if the service is currently running,
+the availability of thee application will not be affected. When you add the
```
Comment (klingerf) [inline linkerd.io/content/2/adding-your-service/_index.md]: Typo here: `thee` => `the`

(or alternatively `thee` => `ye olde`)

File: linkerd.io/content/2/adding-your-service/_index.md [hunk 5]
```diff
@@ -37,8 +51,304 @@ You can always get to the Linkerd dashboard by running:
...
 ```
 
-## Inject Reference
+# Method 2: Automatic Proxy Injection
+
+**Note**: this feature is *experimental*. As of version {{% latestversion %}}, it
+not recommended for production use.
+
+**Note**: Currently, the auto-proxy injection only works with Deployments.
+It doesn't work with other workload like StatefulSet, DaemonSet, Pod etc.
```
Comment (klingerf) [inline linkerd.io/content/2/adding-your-service/_index.md]: Typo here: `workload` => `workloads`

File: linkerd.io/content/2/adding-your-service/_index.md [hunk 6]
```diff
@@ -37,8 +51,304 @@ You can always get to the Linkerd dashboard by running:
...
+
+An alternative approach to adding the Linkerd data plane proxies to services is
```
Comment (klingerf) [inline linkerd.io/content/2/adding-your-service/_index.md]: How about:

> An alternative approach to adding the Linkerd data plane proxies via the command
> line is to instead do it on the cluster, via ...

File: linkerd.io/content/2/adding-your-service/_index.md [hunk 7]
```diff
@@ -37,8 +51,304 @@ You can always get to the Linkerd dashboard by running:
...
+* All required permissions are defined in the new
+  `linkerd-linkerd-proxy-injector` cluster role, separated from the cluster
+  role of the control plane.
+* When a new deployment is created, the Linkerd proxy will be automatically injected if:
+  * the namespace has the `linkerd.io/auto-inject: enabled` label or
+  * the namespace isn't labeled with the `linkerd.io/auto-inject` label.
+* Namespaces that are labeled with the `linkerd.io/auto-inject: disabled` are ignored.
+* Within "enabled" namespaces, deployments with pods that are labeled with
+  `linkerd.io/auto-inject: disabled` or `linkerd.io/auto-inject: completed` are
+  ignored.
```
Comment (klingerf) [inline linkerd.io/content/2/adding-your-service/_index.md]: The formatting here is broken:

image

Adding an extra newline between lines 68 and 69 will fix it.

File: linkerd.io/content/2/adding-your-service/_index.md [hunk 8]
```diff
@@ -37,8 +51,304 @@ You can always get to the Linkerd dashboard by running:
...
+rather than requiring a client-side pipeline during deploys.
+
+Linkerd's automatic proxy injection works as follows:
+* All required permissions are defined in the new
+  `linkerd-linkerd-proxy-injector` cluster role, separated from the cluster
+  role of the control plane.
+* When a new deployment is created, the Linkerd proxy will be automatically injected if:
+  * the namespace has the `linkerd.io/auto-inject: enabled` label or
+  * the namespace isn't labeled with the `linkerd.io/auto-inject` label.
+* Namespaces that are labeled with the `linkerd.io/auto-inject: disabled` are ignored.
```
Comment (klingerf) [inline linkerd.io/content/2/adding-your-service/_index.md]: Missing the word "label" in this bullet point.

> ... are labeled with the `linkerd.io/auto-inject: disabled` _label_ are ignored.

File: linkerd.io/content/2/adding-your-service/_index.md [hunk 9]
```diff
@@ -37,8 +51,304 @@ You can always get to the Linkerd dashboard by running:
...
+    Mounts:       <none>
+
+$ kubectl -n unlabeled get po
+NAME                               READY     STATUS    RESTARTS   AGE
+...
+nginx-enabled-65f8d5c4fb-kh5rr     2/2       Running   0          1m
+```
+The new pod should still have the proxy sidecar.
+
+## Debugging automatic proxy injection.
```
Comment (klingerf) [inline linkerd.io/content/2/adding-your-service/_index.md]: Can you remove the trailing "." character from this title please?

File: linkerd.io/content/2/adding-your-service/_index.md [hunk 10]
```diff
@@ -37,8 +51,304 @@ You can always get to the Linkerd dashboard by running:
...
+proxy-injector logs. If the proxy-injector emits a series of `remote error:
+tls: bad certificate` errors, then the CA bundle defind in the webhook's
+mutating webhook configuration has gone out-of-sync with the CA certificate
+generated by the CA controller. The easiest resolution is to delete the
+proxy-injector pod and let it recreate its mutating webhook configuration.
 
-For more information on how the inject command works and all of the parameters
-that can be set, look at the [reference](../cli/inject/).
+Additionally, the proxy-injector container take the `-log-level` argument which
+can be set to generate more verbose logs output.
```
Comment (klingerf) [inline linkerd.io/content/2/adding-your-service/_index.md]: Let's delete this paragraph about the `-log-level` flag -- that requires modifying the install YAML directly, which I'm not expecting for folks to do.

File: linkerd.io/content/2/architecture/_index.md [hunk 11]
```diff
@@ -3,131 +3,81 @@ date = "2018-09-10T12:00:00-07:00"
 title = "Architecture"
 [menu.l5d2docs]
   name = "Architecture"
-  weight = 2
+  weight = 6
 +++
 
+Linkerd is made of three basic components:
+
+1. The CLI, which runs on your local environment (e.g. your cluster).
```
Comment (klingerf) [inline linkerd.io/content/2/architecture/_index.md]: I don't think the e.g. is correct here -- the CLI doesn't run on your cluster, it runs on, say, your laptop.

File: linkerd.io/content/2/architecture/_index.md [hunk 12]
```diff
@@ -3,131 +3,81 @@ date = "2018-09-10T12:00:00-07:00"
...
+1. The control plane, which runs on your cluster, in a dedicated namespace (e.g. "linkerd").
+1. The data plane, which runs on your cluster, as a set of distributed proxies embedded in your services' pods.
+
+Let's take each of those components in those in turn.
```
Comment (klingerf) [inline linkerd.io/content/2/architecture/_index.md]: Too many "those" in this sentence? Should it be:

> Let's take each of these components in turn.

File: linkerd.io/content/2/architecture/_index.md [hunk 13]
```diff
@@ -3,131 +3,81 @@ date = "2018-09-10T12:00:00-07:00"
...
 
 - Prometheus - All of the metrics exposed by Linkerd are scraped via Prometheus
   and stored here. This is an instance of Prometheus that has been configured to
-  work specifically with the data that Linkerd generates. There are
-  [instructions](/2/observability/prometheus/#exporting-metrics)
-  if you would like to integrate this with an
-  existing Prometheus installation.
+  work specifically with the data that Linkerd generates. *Note*: this
+  Prometheus deployment is intended to supplement, not replace, your existing
+  metrics store. [Read more about how to get metrics out of this Prometheus instance here](/2/observability/prometheus/#exporting-metrics).
```
Comment (klingerf) [inline linkerd.io/content/2/architecture/_index.md]: I don't think you need to include " here" at the end of this link text.

File: linkerd.io/content/2/upgrade/_index.md [hunk 14]
```diff
@@ -23,7 +28,7 @@ these instructions for anywhere that uses the linkerd CLI.
 To upgrade the CLI locally, run:
 
 ```bash
-curl -sL https://run.linkerd.io/install | sh
+curl https://run.linkerd.io/install | sh
```
Comment (grampelberg) [inline linkerd.io/content/2/upgrade/_index.md]: Why'd you remove the `-sL`? It makes the install output cleaner.

File: linkerd.io/content/2/proxy-injection/_index.md [hunk 15]
```diff
@@ -1,306 +0,0 @@
-+++
```
Comment (grampelberg) [inline linkerd.io/content/2/proxy-injection/_index.md]: Where did this content get moved to?

**Comment by grampelberg** [comment]
@klingerf I think this PR breaks some of the links (but I have no idea where they are or how to check it). Any good ideas on some tooling we could add to help out?

**Comment by klingerf** [comment]
@grampelberg Offhand I don't know of an automated way to make sure links are still valid. But you're totally right that we should include aliases for all of the pages being removed.

**Comment by grampelberg** [comment]
@klingerf I'm actually more concerned about the anchor links (which I really have zero ideas how to figure out).

File: linkerd.io/content/2/cli/_index.md [hunk 16]
```diff
@@ -1,6 +1,6 @@
 +++
 date = "2018-09-17T08:00:00-07:00"
-title = "Overview"
+title = "Linkerd CLI"
```
[the remark this answers is not in the document]
Comment (wmorgan) [reply linkerd.io/content/2/cli/_index.md]: It seems to work with the current name, at least locally... are you saying that changing the name break some Hugo thing?

File: linkerd.io/content/2/upgrade/_index.md (hunk 14 above)
Comment (wmorgan) [reply linkerd.io/content/2/upgrade/_index.md]: Was just trying to simplify the command. But we use these flags elsewhere... I'll re-add these in the interest of consistency.

File: linkerd.io/content/2/proxy-injection/_index.md (hunk 15 above)
Comment (wmorgan) [reply linkerd.io/content/2/proxy-injection/_index.md]: The "add your service" page.

File: linkerd.io/content/2/adding-your-service/_index.md (hunk 2 above)
Comment (klingerf) [reply linkerd.io/content/2/adding-your-service/_index.md]: Yep, that works great for me.

File: linkerd.io/content/2/cli/_index.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (grampelberg) [reply linkerd.io/content/2/cli/_index.md]: With `Linkerd CLI` as title:

screenshot 2018-12-03 10 43 48

With `Overview` as title:

screenshot 2018-12-03 10 44 24

File: linkerd.io/content/2/proxy-injection/_index.md (hunk 15 above)
Comment (grampelberg) [reply linkerd.io/content/2/proxy-injection/_index.md]: I don't see anything in there about auto-inject?

File: linkerd.io/content/2/architecture/_index.md [hunk 17]
```diff
@@ -3,131 +3,81 @@ date = "2018-09-10T12:00:00-07:00"
...
-command.
-
-These proxies transparently intercept communication to and from each pod, and
-add features such as instrumentation and encryption (TLS), as well as allowing
-and denying requests according to the relevant policy.
-
-These proxies are not designed to be configured by hand. Rather, their behavior
-is driven by the control plane.
-
-### Proxy
```
[the remark this answers is not in the document]
Comment (grampelberg) [reply linkerd.io/content/2/architecture/_index.md]: The proxy could probably be merged (though there's value in linking to it as a specific component). I'm actually more concerned with removing the dashboard/grafana/prometheus sections. It is pretty nice to be able to link to specific components and have a page that explains exactly what each is and what it does.

**Pull request closed without merging**